flocker/node/test/test_script.py

```diff
@@ -915,3 +920,46 @@ class ContainerAgentOptionsTests(
     """
     Tests for ``ContainerAgentOptions``.
     """
+
+
+class GetExternalIPTests(SynchronousTestCase):
+    """
+    Tests for ``_get_external_ip``.
+    """
+    def setUp(self):
+        server = socket.socket()
+        server.bind(('127.0.0.1', 0))
+        server.listen(5)
+        self.destination_port = server.getsockname()[1]
+        self.addCleanup(server.close)
+
+    @capture_logging(lambda test, logger:
+                     assertHasAction(test, logger, LOG_GET_EXTERNAL_IP, True,
+                                     {u"host": u"localhost",
+                                      u"port": test.destination_port},
+                                     {u"local_ip": u"127.0.0.1"}))
+    def test_successful_get_external_ip(self, logger):
+        """
+        A successful external IP lookup returns the local interface's IP.
+        """
+        self.assertEqual(u"127.0.0.1",
+                         _get_external_ip(u"localhost", self.destination_port))
+
+    @capture_logging(lambda test, logger:
+                     assertHasAction(test, logger, LOG_GET_EXTERNAL_IP, False,
+                                     {u"host": u"localhost",
+                                      u"port": test.destination_port},
+                                     {u"exception":
+                                      u"exceptions.RuntimeError"}))
+    def test_failed_get_external_ip(self, logger):
+        """
+        A failed external IP lookup is retried (and the error logged).
+        """
+        original_connect = socket.socket.connect
+
+        def fail_once(*args, **kwargs):
+            socket.socket.connect = original_connect
+            raise RuntimeError()
+        self.patch(socket.socket, "connect", fail_once)
+        self.assertEqual(u"127.0.0.1",
+                         _get_external_ip(u"localhost", self.destination_port))
```

Comment:
So if faking is good enough here then I think it's definitely good enough for the previous test.